Modules/xASL_module_ASL.m

```diff
@@ -21,10 +21,10 @@
 % A. Check existence and validity of ASL
 % B. Manage mutex state processing step
 % C. Cleanup before rerunning
-% D. ASL processing & quantification parameters
-% E. Backward and forward compatibility of filenames
-% F. Split ASL and M0 within the ASL time series
-% G. DeltaM and CBF parsing - check if all/some volumes are deltams or CBFs
+% D. Backward and forward compatibility of filenames
```

Comment:
Of course, no prob explaining the motivation. I wrote it in the issue:
- ASL split to ASL+M0 happens in BIDS2Legacy, but we still have a failsafe in ASL_module to do this in case people provide data in Legacy themselves. So when ASL+M0 split happens in ASL_module, it needs to happen before ASL parameter check. Because ASL/M0 tend to have different TR/TE/PLD and this would then be falsely identified as a multi-parameter setup. We could have of course done a complicated check in the parameter check, but it's much easier to split before looking for multi-PLD.

Answers to your questions:
1. We didn't do anything else than moving Split ASL forward.
2. DummyScanPosition was not really moved out of ASL_ParseParameters. It was in ASLSplit and stayed there. We only moved the parameter initialization to empty.
3. BASIL quant has to be initialized after bQuantifyMultiPLD as it needs to know this. We've agreed to move bQuantifyMultiPLD to shortly before quantification, so we had to move BASILquant as well.
4. One thing is to calculate parameters, which can be done at ASL module start, but has to be recalculated after merging. Quantification options have to be done before quantification, but has to be run only once to be able to deal with missing parameters and values from dataPar -> so it has to be called only after merging and before quant. So that's why Parameters are called twice, and options once.
5. Parameter function is unconditional at the start. And the recalculation is conditional not for speed-up, but because I need it to run only when merging is done. But I should perhaps run this conditional to merging, but outside of state-check. Same for QuantificationOptions - that has to be run always. Fixed in the last commit.

Reply:
1. OK
2. OK
3. OK
4. OK
5. I thought that we agreed that these parameters were only used for quantification, hence that you would move them to just before quantification (not needing the state checking), and irrespective of merging. Can we remove the merging check? Because I would like the pipeline to do the same irregardless of merging. Merging is a separate function that does merging-related stuff. So I would suggest:
option A: we have to call it twice, then add it to the merging wrapper instead of xASL_module_ASL so the transparency/modularity philosophy that merging doesn't change xASL_module_ASL but is only an insert, if fullfilled
option B: we only have to call this before quantification, then you can remove it at the start and also remove the "if bMerging statement"

6. As per my last comment, it would be great if you can add to the header of the (new) ParseParms functions what is done. So create sections (M0 management, BASIL management, etc), numbering them, and then adding them to the header, like we have in all our functions/wrappers (like here in xASL_module_ASL).

Reply:
5.1 Yes - they are now just before merging and without "state checking".
5.A I can't move it inside the merging function. Because on a re-run, the state will be finished and merging won't be re-run.
5.B I don't want to remove it at the start because we want to have access to UniqueParameter vectors potentially even before quantification.

The pipeline of course does the same irregardless of the merging. If we run the MultiParam function again without merging, it will have 100% same output and zero effect. I would really prefer to not run things twice with the only reason that the code looks nice. It will have zero influence to run it the second time, it will just consume a few extra seconds to read it again the file etc. It's a really bad programming practice. But if you really insist, I can do that.

6. Yes. Added

Reply:
5. Okay, compromise, you can add it as elseif to the if-statement of the merging state control right?

If State{notyetmerged}
     Merging function
    % we need to reinitialize the parameters from the merged PWI4D.json
     ParseParameters
elseif bMerge
    % we need to reinitialize the parameters from the merged PWI4D.json
    ParseParameters
end

or at least just keep it in the same code section, that it is clear that this is "an insert".

But I understand now why you need to do it like this.
I would still say that if it costs 0 time, it might be bad programming practice but clearer since we will always do the same, i.e. rechecking the quantification parameters separately for the quantification compared to the processing.

Thinking about this, isn't this more logical:

1. ASL4D.json provides the settings/data for how we should process our data
2. PWI4D.json provides settings/data for how we should quantify our data (can be from ASL4D.json but can also be either manually or automatically merged)
3. xASL_wrp_Quantify re-reads both PWI4D.nii AND PWI4D.json. Because we actually don't know if the user has manually merged two NIfTIs!

Even if you disagree that the user should be able to manually merge two NIfTIs (which is transparent and doable now), this is even more clear. And we don't need to say "if merging, we decide that the quantification parameters potentially differ from the processing parameters, but if no merging, we are certain that this is not the case".

So I would propose doing this for simplification and transparency.